Crash in GeoJSonUtils on opening the map

User on support reports the map is always crashing for her. Error log points to GeoJsonUtils, see below:

```
io.reactivex.rxjava3.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | java.lang.OutOfMemoryError: Failed to allocate a 268959752 byte allocation with 16777216 free bytes and 228MB until OOM
at io.reactivex.rxjava3.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:372)
at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:68)
at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
at java.util.concurrent.FutureTask.run(FutureTask.java:237)
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
at java.lang.Thread.run(Thread.java:760)
Caused by: java.lang.OutOfMemoryError: Failed to allocate a 268959752 byte allocation with 16777216 free bytes and 228MB until OOM
at java.util.Arrays.copyOf(Arrays.java:3352)
at java.lang.AbstractStringBuilder.expandCapacity(AbstractStringBuilder.java:130)
at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:114)
at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:550)
at java.lang.StringBuilder.append(StringBuilder.java:205)
at com.cocoahero.android.geojson.util.StreamUtils.toString(StreamUtils.java:38)
at com.cocoahero.android.geojson.util.StreamUtils.toString(StreamUtils.java:18)
at com.cocoahero.android.geojson.GeoJSON.parse(GeoJSON.java:135)
at cgeo.geocaching.utils.GeoJsonUtils.parseGeoJson(GeoJsonUtils.java:57)
at cgeo.geocaching.files.GPXTrackOrRouteImporter.parseAsGeoJson(GPXTrackOrRouteImporter.java:120)
at cgeo.geocaching.files.GPXTrackOrRouteImporter.doInBackground(GPXTrackOrRouteImporter.java:89)
at cgeo.geocaching.files.GPXTrackOrRouteImporter.lambda$doImport$1(GPXTrackOrRouteImporter.java:39)
at cgeo.geocaching.files.GPXTrackOrRouteImporter.$r8$lambda$g0pSx27yHp6Zq7CsAgfCAE4gnXU(GPXTrackOrRouteImporter.java)
at cgeo.geocaching.files.GPXTrackOrRouteImporter$$ExternalSyntheticLambda0.run(R8$$SyntheticClass)
at cgeo.geocaching.utils.AndroidRxUtils.lambda$andThenOnUi$2(AndroidRxUtils.java:91)
at cgeo.geocaching.utils.AndroidRxUtils.$r8$lambda$bqzWo9nlvU86BvboB-9jhDWp5vc(AndroidRxUtils.java)
at cgeo.geocaching.utils.AndroidRxUtils$$ExternalSyntheticLambda0.run(R8$$SyntheticClass)
at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
... 6 more
```
Release 2023.02.12
ticket 724300

Edit: I asked the user whether she has some tracks activated on map and whether she can send us the files so that we can debug into it.

We have another case of this now on support:

User reports, that since upgrade to the new version it crashes each time he wants to open the map.
Pretty sure he did not use any Geojson yet.


```
02-16 23:28:08.184 13862 14316 E AndroidRuntime: FATAL EXCEPTION: RxCachedThreadScheduler-4
02-16 23:28:08.184 13862 14316 E AndroidRuntime: Process: cgeo.geocaching, PID: 13862
02-16 23:28:08.184 13862 14316 E AndroidRuntime: io.reactivex.rxjava3.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | java.lang.OutOfMemoryError: Failed to allocate a 349693734 byte allocation with 16777216 free bytes and 172MB until OOM
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at io.reactivex.rxjava3.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:372)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:68)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:154)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:269)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:818)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: Caused by: java.lang.OutOfMemoryError: Failed to allocate a 349693734 byte allocation with 16777216 free bytes and 172MB until OOM
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at java.lang.AbstractStringBuilder.enlargeBuffer(AbstractStringBuilder.java:95)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at java.lang.AbstractStringBuilder.append0(AbstractStringBuilder.java:125)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at java.lang.StringBuilder.append(StringBuilder.java:271)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at com.cocoahero.android.geojson.util.StreamUtils.toString(StreamUtils.java:38)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at com.cocoahero.android.geojson.util.StreamUtils.toString(StreamUtils.java:18)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at com.cocoahero.android.geojson.GeoJSON.parse(GeoJSON.java:135)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at cgeo.geocaching.utils.GeoJsonUtils.parseGeoJson(GeoJsonUtils.java:57)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at cgeo.geocaching.files.GPXTrackOrRouteImporter.parseAsGeoJson(GPXTrackOrRouteImporter.java:120)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at cgeo.geocaching.files.GPXTrackOrRouteImporter.doInBackground(GPXTrackOrRouteImporter.java:89)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at cgeo.geocaching.files.GPXTrackOrRouteImporter.lambda$doImport$1(GPXTrackOrRouteImporter.java:39)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at cgeo.geocaching.files.GPXTrackOrRouteImporter.$r8$lambda$g0pSx27yHp6Zq7CsAgfCAE4gnXU(GPXTrackOrRouteImporter.java)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at cgeo.geocaching.files.GPXTrackOrRouteImporter$$ExternalSyntheticLambda0.run(R8$$SyntheticClass)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at cgeo.geocaching.utils.AndroidRxUtils.lambda$andThenOnUi$2(AndroidRxUtils.java:91)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at cgeo.geocaching.utils.AndroidRxUtils.$r8$lambda$bqzWo9nlvU86BvboB-9jhDWp5vc(AndroidRxUtils.java)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at cgeo.geocaching.utils.AndroidRxUtils$$ExternalSyntheticLambda0.run(R8$$SyntheticClass)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
02-16 23:28:08.184 13862 14316 E AndroidRuntime: 	... 7 more
```

--- a/main/src/main/java/cgeo/geocaching/files/GPXTrackOrRouteImporter.java
+++ b/main/src/main/java/cgeo/geocaching/files/GPXTrackOrRouteImporter.java
@@ -7,6 +7,8 @@ import cgeo.geocaching.location.IGeoDataProvider;
 import cgeo.geocaching.models.Route;
 import cgeo.geocaching.storage.ContentStorage;
 import cgeo.geocaching.utils.AndroidRxUtils;
+import cgeo.geocaching.utils.EnvironmentUtils;
+import cgeo.geocaching.utils.Formatter;
 import cgeo.geocaching.utils.GeoJsonUtils;
 import cgeo.geocaching.utils.Log;
 
@@ -36,7 +38,7 @@ public class GPXTrackOrRouteImporter {
         final AtomicBoolean success = new AtomicBoolean(false);
         AndroidRxUtils.andThenOnUi(Schedulers.io(), () -> {
             try {
-                final IGeoDataProvider value = doInBackground(uri);
+                final IGeoDataProvider value = doInBackground(context, uri);
                 success.set(null != value && value.hasData());
                 if (success.get()) {
                     AndroidSchedulers.mainThread().createWorker().schedule(() -> {
@@ -58,7 +60,9 @@ public class GPXTrackOrRouteImporter {
         });
     }
 
-    private static IGeoDataProvider doInBackground(final Uri uri) {
+    // splitting up that method would not help improve readability
+    @SuppressWarnings({"PMD.NPathComplexity", "PMD.ExcessiveMethodLength"})
+    private static IGeoDataProvider doInBackground(final Context context, final Uri uri) {
         try {
             // default: import properly formatted routes or tracks
             Route route = parse(new GPXTrackParser("http://www.topografix.com/GPX/1/1", "1.1"), uri);
@@ -86,7 +90,7 @@ public class GPXTrackOrRouteImporter {
                 route.calculateNavigationRoute();
             }
             if (null == route) {
-                return parseAsGeoJson(uri);
+                return parseAsGeoJson(context, uri);
             }
             return route;
         } catch (IOException e) {
@@ -112,7 +116,15 @@ public class GPXTrackOrRouteImporter {
         }
     }
 
-    private static IGeoDataProvider parseAsGeoJson(final Uri uri) throws IOException {
+    private static IGeoDataProvider parseAsGeoJson(final Context context, final Uri uri) throws IOException {
+        final ContentStorage.FileInformation fi = ContentStorage.get().getFileInfo(uri);
+        final long freeMem = EnvironmentUtils.getFreeMemory(context);
+        if (fi == null || freeMem < 0 || fi.size * 10 > freeMem) {
+            Log.w("Won't import '" + uri + "' as json due to limited memory (filesize: " +
+                    Formatter.formatBytes(fi == null ? 0 : fi.size) + ", freeMem: " + Formatter.formatBytes(freeMem));
+            return null;
+        }
+
         try (InputStream is = ContentStorage.get().openForRead(uri)) {
             if (is == null) {
                 return null;

--- a/main/src/main/java/cgeo/geocaching/utils/EnvironmentUtils.java
+++ b/main/src/main/java/cgeo/geocaching/utils/EnvironmentUtils.java
@@ -1,7 +1,12 @@
 package cgeo.geocaching.utils;
 
+import android.app.ActivityManager;
+import android.content.Context;
 import android.os.Build;
 import android.os.Environment;
+import static android.content.Context.ACTIVITY_SERVICE;
+
+import androidx.annotation.Nullable;
 
 public class EnvironmentUtils {
     private static final boolean DEFAULT_TRANSPARENT_BACKGROUND;
@@ -24,5 +29,25 @@ public class EnvironmentUtils {
 
     public static boolean defaultBackgroundTransparent() {
         return DEFAULT_TRANSPARENT_BACKGROUND;
+    }
+
+    @Nullable
+    public static ActivityManager.MemoryInfo getMemoryInfo(final Context context) {
+        if (context == null) {
+            return null;
+        }
+        final ActivityManager activityManager = (ActivityManager) context.getSystemService(ACTIVITY_SERVICE);
+        if (activityManager == null) {
+            return null;
+        }
+        final ActivityManager.MemoryInfo memoryInfo = new ActivityManager.MemoryInfo();
+        activityManager.getMemoryInfo(memoryInfo);
+        return memoryInfo;
+    }
+
+    public static long getFreeMemory(final Context context) {
+        final ActivityManager.MemoryInfo memInfo = getMemoryInfo(context);
+        return memInfo == null ? -1 : memInfo.availMem;
+
     }
 }

--- a/main/src/main/java/cgeo/geocaching/utils/SystemInformation.java
+++ b/main/src/main/java/cgeo/geocaching/utils/SystemInformation.java
@@ -25,6 +25,7 @@ import cgeo.geocaching.storage.LocalStorage;
 import cgeo.geocaching.storage.PersistableFolder;
 import cgeo.geocaching.storage.PersistableUri;
 
+import android.app.ActivityManager;
 import android.content.Context;
 import android.content.UriPermission;
 import android.content.pm.PackageManager;
@@ -79,6 +80,7 @@ public final class SystemInformation {
         appendScreenResolution(context, body);
         body.append("\n- Sailfish OS detected: ").append(EnvironmentUtils.isSailfishOs());
         appendGooglePlayServicesVersion(context, body);
+        appendMemoryInfo(context, body);
         body.append("\n")
                 .append("\nSensor and location:")
                 .append("\n-------")
@@ -147,6 +149,19 @@ public final class SystemInformation {
 
         body.append("\n\n--- End of system information ---\n");
         return body.toString();
+    }
+
+    private static void appendMemoryInfo(@NonNull final Context context, @NonNull final StringBuilder body) {
+        body.append("\n- Memory: ");
+        final ActivityManager.MemoryInfo memoryInfo = EnvironmentUtils.getMemoryInfo(context);
+        if (memoryInfo == null) {
+            body.append("null");
+        } else {
+            body.append(" Available:" + Formatter.formatBytes(memoryInfo.availMem) +
+                    ", Total:" + Formatter.formatBytes(memoryInfo.totalMem) +
+                    ", Threshold: " + Formatter.formatBytes(memoryInfo.threshold) +
+                    ", low:" + memoryInfo.lowMemory);
+        }
     }
 
     private static void appendDatabase(@NonNull final StringBuilder body) {

